src/DataStreams/RemoteQueryExecutor.cpp

```diff
@@ -191,7 +201,37 @@ Block RemoteQueryExecutor::read()
         if (context.getSettingsRef().skip_unavailable_shards && (0 == multiplexed_connections->size()))
             return {};
     }
+    try
+    {
+        return readPackets();
+    }
+    catch (const DB::Exception & ex)
+    {
+        if (ex.code() == ErrorCodes::DUPLICATED_PART_UUIDS)
+        {
+            /// Cancel previous query and disconnect before retry.
+            cancel();
+            multiplexed_connections->disconnect();
+
+            /// Only resend once, otherwise throw an exception
```

Comment:
Am I understand correctly that resend this only once is not enough? Since there is no barrier for the `PartUuids` package (i.e. remote node does not wait until all packages from other nodes will be received)

Never mind, looks like this was already discussed - https://github.com/ClickHouse/ClickHouse/pull/17348#discussion_r529467372

Reply:
Also seems that this will not work with `distributed_group_by_no_merge`? (Worth a comment in this case)

Reply:
it works with `distributed_group_by_no_merge=1`

Reply:
>it works with distributed_group_by_no_merge=1

OK, I see, this is because `Protocol::Server::PartUUIDs` sent before data blocks.